algorithms/src/snark/groth16/r1cs_to_qap.rs

```diff
@@ -102,50 +94,33 @@ impl R1CStoQAP {
 
     #[inline]
     pub(crate) fn witness_map<E: PairingEngine>(prover: &ProvingAssignment<E>) -> SynthesisResult<Vec<E::Fr>> {
-        let zero = E::Fr::zero();
         let num_inputs = prover.public_variables.len();
         let num_constraints = prover.num_constraints();
 
         let full_input_assignment = [&prover.public_variables[..], &prover.private_variables[..]].concat();
 
         let domain = EvaluationDomain::<E::Fr>::new(num_constraints + num_inputs)
             .ok_or(SynthesisError::PolynomialDegreeTooLarge)?;
-        let domain_size = domain.size();
 
-        let mut a = vec![zero; domain_size];
-        let mut b = vec![zero; domain_size];
+        let mut evaluated = cfg_into_iter!([&prover.at, &prover.bt, &prover.ct])
+            .map(|cstr| evaluate_constraints::<E>(cstr, &full_input_assignment, &domain, num_constraints, num_inputs))
+            .collect::<Vec<_>>();
 
-        cfg_iter_mut!(a[..num_constraints])
-            .zip(cfg_iter_mut!(b[..num_constraints]))
-            .zip(cfg_iter!(&prover.at))
-            .zip(cfg_iter!(&prover.bt))
-            .for_each(|(((a, b), at_i), bt_i)| {
-                *a = evaluate_constraint::<E>(at_i, &full_input_assignment, num_inputs);
-                *b = evaluate_constraint::<E>(bt_i, &full_input_assignment, num_inputs);
-            });
+        let mut c = evaluated.pop().unwrap();
+        let mut b = evaluated.pop().unwrap();
+        let mut a = evaluated.pop().unwrap();
 
         a[num_constraints..(num_inputs + num_constraints)].clone_from_slice(&full_input_assignment[..num_inputs]);
 
-        domain.ifft_in_place(&mut a);
-        domain.ifft_in_place(&mut b);
-
-        domain.coset_fft_in_place(&mut a);
-        domain.coset_fft_in_place(&mut b);
+        cfg_into_iter!([&mut a, &mut b, &mut c]).for_each(|cstr| {
+            domain.ifft_in_place(cstr);
+            domain.coset_fft_in_place(cstr);
+        });
```

Comment:
Maybe for an additional speedup we can use the `ExecutionPool` here to run all 6 jobs in parallel, instead of running 3 groups of 2 in parallel?

Reply:
Hm doesn't that line already complete before this line?

Reply:
ah, I thought you meant to include the previous operations; I can check if using `ExecutionPool` here would result in greater performance :+1:

Reply:
if this is what you had in mind:
```
--- a/algorithms/src/snark/groth16/r1cs_to_qap.rs
+++ b/algorithms/src/snark/groth16/r1cs_to_qap.rs
@@ -106,16 +106,26 @@ impl R1CStoQAP {
             .map(|cstr| evaluate_constraints::<E>(cstr, &full_input_assignment, &domain, num_constraints, num_inputs))
             .collect::<Vec<_>>();
 
-        let mut c = evaluated.pop().unwrap();
-        let mut b = evaluated.pop().unwrap();
+        let c = evaluated.pop().unwrap();
+        let b = evaluated.pop().unwrap();
         let mut a = evaluated.pop().unwrap();
 
         a[num_constraints..(num_inputs + num_constraints)].clone_from_slice(&full_input_assignment[..num_inputs]);
 
-        cfg_into_iter!([&mut a, &mut b, &mut c]).for_each(|cstr| {
-            domain.ifft_in_place(cstr);
-            domain.coset_fft_in_place(cstr);
-        });
+        let mut pool = snarkvm_utilities::ExecutionPool::<Vec<<E as PairingEngine>::Fr>>::with_capacity(3);
+
+        for mut cstr in [a, b, c] {
+            pool.add_job(move || {
+                domain.ifft_in_place(&mut cstr);
+                domain.coset_fft_in_place(&mut cstr);
+                cstr
+            });
+        }
+
+        let mut result = pool.execute_all();
+        let c = result.pop().unwrap();
+        let b = result.pop().unwrap();
+        let a = result.pop().unwrap();
 
         let mut ab = domain.mul_polynomials_in_evaluation_domain(&a, &b);
         drop(a);
```
such change results in 40% worse performance than the current shape of this PR

Reply:
Ah the execution pool use the maximum number of physical threads https://github.com/AleoHQ/snarkVM/blob/51633e2fac247bc829bd316a93f59b6ba0759661/utilities/src/parallel.rs#L56

So it basically undoes the removal of the `execute_with_max_available_threads` in the `fft` module.

@howardwu this warrants further investigation; @ljedrz what kind of machine are you using? Howard saw throttling on his 28-core Intel Mac Pro when using hyperthreads, but I didn't see any throttling on my 24-core (48-hyperthread) Intel machine.

Reply:
my CPU is AMD Ryzen 9 3900X 12-Core (2 threads per core).